src/aiida_quantumespresso/parsers/xspectra.py

```diff
@@ -0,0 +1,205 @@
+        from aiida.plugins import DataFactory
+        
+        retrieved = self.retrieved
+        try:
+            filename_stdout = self.node.get_option('output_filename')
+            with retrieved.open(filename_stdout, 'r') as fil:
+                out_file = fil.readlines()
+        except OSError:
+            return self.exit(self.exit_codes.ERROR_OUTPUT_STDOUT_READ)
+
+        job_done = False
+        for i in range(len(out_file)):
+            line = out_file[-i]
+            if 'Wrong xiabs!!!' in line:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_ABSORBING_SPECIES_WRONG)
+            if 'xiabs < 1 or xiabs > ntyp' in line:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_ABSORBING_SPECIES_ZERO)
+            if 'END JOB' in line:
+                job_done = True
+                break
+        if not job_done:
+            return self.exit(self.exit_codes.ERROR_OUTPUT_STDOUT_INCOMPLETE)
+        
+        try:
+            with retrieved.open(self.node.process_class._Spectrum_FILENAME, 'r') as fil:
+                xspectra_file = fil.readlines()
+        except OSError:
+            return self.exit(self.exit_codes.ERROR_READING_XSPECTRA_FILE)
+        
+        # end of initial checks
+        
+        array_names = ['energy', 'calculated_intensity']
+        array_units = ['eV', 'n/a']
+        
+        array_data = parse_raw_xspectra(xspectra_file, array_names, array_units)
+        
+        energy_units = 'eV'
+        xy_data = XyData()
+        xy_data.set_x(array_data['energy'], 'energy', energy_units)
+        
+        y_arrays = []
+        y_names = []
+        y_units = []
+        y_arrays += [array_data['calculated_intensity']]
+        y_names += ['calculated_intensity']
+        y_units += ['n/a']
+        xy_data.set_y(y_arrays, y_names, y_units)
+
+        parsed_data, logs = parse_stdout_xspectra(filecontent=out_file, codename='XSpectra')
+        self.emit_logs(logs)
+
+        self.out('spectra_data', xy_data)
+        self.out('output_parameters', Dict(dict=parsed_data))
+        
+def parse_raw_xspectra(xspectra_file, array_names, array_units):
+    """This function takes as input the xspectra_file as a list of filelines 
+    along with information on how to give labels and units to the parsed data.
+
+    :param xspectra_file: xspectra file lines in the form of a list
+    :type xspectra_file: list
+    :param array_names: list of all array names.
+    :type array_names: list
+    :param array_units: list of all array units.
+    :type array_units: list
+
+    :return array_data: narray, a dictionary for ArrayData type, which 
+    contains all parsed xspectra output along with labels and units
+    """
+
+    xspectra_header = xspectra_file[:4]
+    try:
+        xspectra_data = np.genfromtxt(xspectra_file)
+    except ValueError:
+        raise QEOutputParsingError('XSpectra file could not be loaded using genfromtxt')
```

Comment:
You are not catching these exceptions right? Would probably be best to catch these in the `parse` method and convert to an exit code instead of letting the parser except.